crates/wash-lib/src/plugin/subcommand.rs

```diff
@@ -0,0 +1,188 @@
+        let ctx = WasiCtxBuilder::new().build();
+
+        let ctx = Data {
+            table: wasmtime::component::ResourceTable::default(),
+            ctx,
+            http: WasiHttpCtx,
+        };
+
+        let mut store = wasmtime::Store::new(&self.engine, ctx);
+
+        let component = Component::from_file(&self.engine, &path)?;
+        let mut linker = Linker::new(&self.engine);
+        wasmtime_wasi::command::add_to_linker(&mut linker)?;
+        wasmtime_wasi_http::bindings::http::outgoing_handler::add_to_linker(
+            &mut linker,
+            |state: &mut Data| state,
+        )?;
+        wasmtime_wasi_http::bindings::http::types::add_to_linker(
+            &mut linker,
+            |state: &mut Data| state,
+        )?;
+
+        let (instance, _) = Subcommands::instantiate_async(&mut store, &component, &linker).await?;
+
+        let metadata = instance.interface0.call_register(&mut store).await?;
+        let maybe_existing = self.plugins.insert(
+            metadata.id.clone(),
+            InstanceData {
+                instance,
+                metadata: metadata.clone(),
+                loaded_path: path.as_ref().to_owned(),
+                store,
+            },
+        );
+
+        match (update, maybe_existing) {
+            // If we're updating and the plugin exists already, overwrite is ok.
+            (true, _) | (false, None) => Ok(metadata),
+            // If update isn't set, then we don't allow the update
+            (false, Some(plugin)) => {
+                // Insert the existing plugin back into the map
+                let id = plugin.metadata.id.clone();
+                self.plugins.insert(plugin.metadata.id.clone(), plugin);
+                Err(anyhow::anyhow!("Plugin with id {id} already exists"))
+            }
+        }
+    }
+
+    /// Get the metadata for a plugin with the given ID if it exists.
+    pub fn metadata(&self, id: &str) -> Option<&Metadata> {
+        self.plugins.get(id).map(|p| &p.metadata)
+    }
+
+    /// Returns a list of all metadata for all plugins.
+    pub fn all_metadata(&self) -> Vec<&Metadata> {
+        self.plugins.values().map(|data| &data.metadata).collect()
+    }
+
+    /// Returns the path to the plugin with the given ID.
+    pub fn path(&self, id: &str) -> Option<&Path> {
+        self.plugins.get(id).map(|p| p.loaded_path.as_path())
+    }
+
+    /// Run a subcommand with the given name and args. The plugin will inherit all
+    /// stdout/stderr/stdin/env. The given plugin_dir is used to grant the plugin access to the
+    /// filesystem in a specific directory, and should already exist. An error will only be returned
+    /// if there was a problem with the plugin (such as the plugin_dir not existing) or the
+    /// subcommand itself.
+    ///
+    /// All plugins will be passed environment variables starting with
+    /// `WASH_PLUGIN_${plugin_id.to_upper()}_` from the current process. Other vars will be ignored
```

Comment:
This will be good to add to documentation since it's likely folks will just expect the regular variable. I think this is a good measure to not give all plugins all-access

Reply:
Yeah all of this will be added to documentation